File: 3585.find-weighted-median-node-in-tree.py
#
# @lc app=leetcode id=3585 lang=python3
#
# [3585] Find Weighted Median Node in Tree
#

# @lc code=start
class Solution:
    def findMedian(self, n: int, edges: List[List[int]], queries: List[List[int]]) -> List[int]:
        from collections import defaultdict, deque
        
        # Step 1: Create an adjacency list for the tree
        adj_list = defaultdict(list)
        for u, v, w in edges:
            adj_list[u].append((v, w))
            adj_list[v].append((u, w))
        
        # Step 2: Precompute all paths using BFS for efficient querying
        def bfs(root):
            parent = {root: None}
            depth = {root: 0}
            queue = deque([root])
            while queue:
                node = queue.popleft()
                for neighbor, weight in adj_list[node]:
                    if neighbor not in parent:
                        parent[neighbor] = (node, weight)
                        depth[neighbor] = depth[node] + 1
                        queue.append(neighbor)
            return parent, depth
        
        # Precompute parent and depth information from root node (0)
        parent_info, depth_info = bfs(0)
        
        # Function to find LCA of two nodes using parent information and depths
        def find_lca(u, v):
            if depth_info[u] < depth_info[v]:
                u, v = v, u
            
            while depth_info[u] > depth_info[v]:
                u = parent_info[u][0]
            
            while u != v:
                u = parent_info[u][0]
                v = parent_info[v][0]
            
            return u
        
        # Function to find weighted median given a path from u to v via LCA lca
        def find_weighted_median(u, v):
            lca = find_lca(u, v)
            path_weight = 0
            weights_u_to_lca = []  # Weights from u to LCA inclusive
            weights_v_to_lca = []  # Weights from v to LCA exclusive (reverse order)									# Ascend from u to lca storing weights along path.	while u != lca:	parent_node_u , weight_u_via_parent= parent_info[u]	weights_u_to_lca.append(weight_u_via_parent) 	u=parent_node_u continue until reach common ancestor node (lca).	Ascend similarly from query end vertex(v) towards lca storing edge-weight(s) encountered until reach this shared ancestor point between both vertices(v,u).	while(v!=lca):	parent_node_v , weight_v_via_parent=parent_info[v]	weights_v_to_lca.append(weight_v_via_parent) # Note : Reverse order due descending upwards towards ancestor before reversing again when calculating cumulative weight sums later on . This enables easier summation calculation(s).	v=parent_node_v continue until meet common intermediary node shared with previous traversal upwards prior within function call stack .# Total path weight is sum of both traversals' cumulative weights :path_weight=sum(weights_u_to_lca)+sum(weights_v_to_lca)# Calculate half-path-weight threshold value we seek surpassing/meeting during traversal summation(s):half_weight_threshold=path_weight/2.0current_sum_weight=0.0for idx,node_weight in enumerate(weights_u_to_lca):current_sum_weight+=node_weightif current_sum_weight>=half_weight_threshold:return idx+1if not idx+1==len(weights_u_to_lca)-1 else [u,lca][idx==len(weights_u_to_lca)-1]for node_wgt_idx,node_wgt_rvs in enumerate(reversed(weights_v_to_lca)):current_sum_weight+=node_wgt_rvsif current_sum_weight>=half_weight_threshold:return [lca,v][node_wgt_idx==len(reversed(weights_v_to_lca))-1]return-1# should not hit under constraints .    results=[]for query_start_vertex ,query_end_vertex in queries :results.append(find_weighted_median(query_start_vertex ,query_end_vertex))      return results

File: 3585.find-weighted-median-node-in-tree.cpp
#
# @lc app=leetcode id=3585 lang=cpp
#
# [3585] Find Weighted Median Node in Tree
#

# @lc code=start
class Solution {
public:
    static const int LOG = 20;
    vector<vector<pair<int, int>>> adj;
    vector<vector<int>> up; // up[v][k] = 2^k ancestor of v
    vector<int> depth;
    vector<long long> preSum; // prefix sum of edge weights from root

    void dfs(int v, int p, int d, long long sum) {
        up[v][0] = p;
        depth[v] = d;
        preSum[v] = sum;
        for (auto &[to, w] : adj[v]) {
            if (to != p) dfs(to, v, d + 1, sum + w);
        }
    }

    int lca(int u, int v) {
        if (depth[u] < depth[v]) std::swap(u, v);
        for (int k = LOG - 1; k >= 0; --k) {
            if (depth[u] - (1 << k) >= depth[v]) u = up[u][k];
        }
        if (u == v) return u;
        for (int k = LOG - 1; k >= 0; --k) {
            if (up[u][k] != up[v][k]) {
                u = up[u][k];
                v = up[v][k];
            }
        }
        return up[u][0];
    }

    int go_up(int u, int steps) {
        for (int k = 0; k < LOG; ++k) {
            if (steps & (1 << k)) u = up[u][k];
        }
        return u;
    }

    vector<int> findMedian(int n, vector<vector<int>>& edges, vector<vector<int>>& queries) {
        adj.assign(n, {});
        for (auto &e : edges) {
            int u = e[0], v = e[1], w = e[2];
            adj[u].emplace_back(v, w);
            adj[v].emplace_back(u, w);
        }
        up.assign(n, vector<int>(LOG));
        depth.assign(n, 0);
        preSum.assign(n, 0);
        dfs(0, 0, 0, 0);
        for (int k = 1; k < LOG; ++k) {
            for (int v = 0; v < n; ++v) {
                up[v][k] = up[up[v][k - 1]][k - 1];
            }
        }
        vector<int> ans;
        for (auto &q : queries) {
            int u = q[0], v = q[1];
            int anc = lca(u, v);
            long long total = preSum[u] + preSum[v] - 2 * preSum[anc];
            long long half = (total + 1) / 2; // strictly at least half
            // Go from u towards v
            int cur = u;
            long long sum = 0;
            if (u == v) {
                ans.push_back(u);
                continue;
            }
            // First, try to go up from u to anc
            for (int k = LOG - 1; k >= 0; --k) {
                if (depth[cur] - (1 << k) >= depth[anc]) {
                    int nxt = up[cur][k];
                    long long seg = preSum[cur] - preSum[nxt];
                    if (sum + seg >= half) {
                        // need to binary search down this segment
                        int low = cur, high = nxt, res = cur;
                        while (low != high) {
                            int mid = go_up(low, depth[low] - depth[high] - 1);
                            if (sum + preSum[cur] - preSum[mid] >= half) low = mid;
                            else { res = mid; high = mid; }
                        }
                        ans.push_back(low);
                        goto next_query;
                    } else {
                        sum += seg;
                        cur = nxt;
                    }
                }
            }
            // Now go down from anc to v
            vector<int> path;
            int x = v;
            while (x != anc) { path.push_back(x); x = up[x][0]; }
            reverse(path.begin(), path.end());
            for (int node : path) {
                long long seg = preSum[node] - preSum[up[node][0]];
                if (sum + seg >= half) {
                    ans.push_back(node);
                    goto next_query;
                } else {
                    sum += seg;
                }
            }
            ans.push_back(v);
        next_query:;
        }
        return ans;
    }
};
# @lc code=end